marlowe-spec-test/src/Marlowe/Spec/Core/Guarantees.hs

```diff
@@ -0,0 +1,324 @@
+--        isQuiescent (txOutContract traOut) (txOutState traOut)"
+-- TODO: Improve test-data quality. Less than 10% of txinput are empty transaction, and manually checking I noticed a lot of timeouted contracts
+computeTransactionIsQuiescentTest :: InterpretJsonRequest -> TestTree
+computeTransactionIsQuiescentTest interpret = reproducibleProperty "theorem computeTransactionIsQuiescent" do
+    context <- run $ generateT $ genContext interpret
+    (contract, state, transactions) <- run $ generate $
+        frequency
+          [ (45, genContractStateAndValidTransactions context `suchThat` \(_,_,l) -> not (null l))
+          , (55, genContractStateAndValidTransactions context `suchThat` \(_,_,l) -> length l > 1)
+          ]
+    let
+        transaction = head transactions
+        req :: Request JSON.Value
+        req = ComputeTransaction transaction state contract
+
+    RequestResponse res <- run $ liftIO $ interpret req
+    case JSON.fromJSON res of
+      JSON.Success (TransactionOutput o) -> do
+        monitor
+          ( counterexample $
+              "Request: " ++ showAsJson req ++ "\n"
+                ++ "Response: " ++ showAsJson res ++ "\n"
+                ++ "Expected reponse to be quiescent")
+        stop $ cover 30.0 (not (null (inputs transaction))) "Non empty transaction"
+             $ cover 10.0 (isWhen $ txOutContract o) "Output contract is a When statement"
+             $ isQuiescent (txOutContract o) (txOutState o)
+             :: PropertyM ReproducibleTest Bool
+      JSON.Success (TransactionError err ) -> fail $ "Unexpected Transaction Error: " ++ show err
+      _ -> fail "JSON parsing failed!"
+
+
+
+-- QuiescentResults.thy
+--
+-- theorem playTraceIsQuiescent:
+--    "playTrace sl cont (Cons h t) = TransactionOutput traOut ⟹
+--      isQuiescent (txOutContract traOut) (txOutState traOut)"
+playTraceIsQuiescentTest :: InterpretJsonRequest -> TestTree
+playTraceIsQuiescentTest interpret = reproducibleProperty "theorem playTraceIsQuiescent" do
+    context <- run $ generateT $ genContext interpret
+    (contract, State_ext _ _ _ (Arith.Int_of_integer startTime) _, transactions) <- run $ generate $
+        frequency
+          [ (45, genContractStateAndValidTransactions context `suchThat` \(_,_,l) -> not (null l))
+          , (55, genContractStateAndValidTransactions context `suchThat` \(_,_,l) -> length l > 1)
+          ]
+    let
+        req :: Request JSON.Value
+        req = PlayTrace startTime contract transactions
+    RequestResponse res <- run $ liftIO $ interpret req
+
+    case JSON.fromJSON res of
+      JSON.Success (TransactionOutput o) -> do
+        monitor
+          ( counterexample $
+              "Request: " ++ showAsJson req ++ "\n"
+                ++ "Response: " ++ showAsJson res ++ "\n"
+                ++ "Expected reponse to be quiescent" )
+        assert $ isQuiescent (txOutContract o) (txOutState o)
+      JSON.Success (TransactionError _ ) -> pre False
```

Comment:
This can obscure a potential error. Tests should be discarded at generation time, not after making the test if they failed